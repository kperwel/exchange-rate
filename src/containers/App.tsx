import React, { useState } from "react";
import CurrencyAwareMoneyInput from "../components/CurrencyAwareMoneyInput/CurrencyAwareMoneyInput";
import Button from "../components/Button/Button";
import Slider from "../components/Slider/Slider";
import useCurrencies, { STATUS } from "../hooks/useCurrencies";
import { RateValue } from "../types";
import styled from "styled-components";

const AppStyled = styled.div`
  display: flex;
  flex-direction: column;
  background: #fff;
  width: 100%;
  min-height: 100%;
  background-image: linear-gradient(120deg, #e0c3fc 0%, #8ec5fc 100%);
  
  justify-content: center;
  align-items: center;
`;

const InputSectionStyled = styled.div`
  background: #fff;
  max-width: 100%;
  display: flex;
  border-radius: 10px;
  flex-direction: column;
  padding: 10px;
  margin: 5px;
`;

const MenuSectionStyled = styled.div`
  background: #fff;
  max-width: 100%;
  display: flex;
  justify-items: space-between;
  flex-direction: row;
`;

const CurrencySelectSection = styled.div`
  background: rgba(255, 255, 255, 0.2);
`;

// Second parameter for round method is a RoundingMode
// Sadly TS with isolated modules do not allow to use
// exported const enums from external module
const ROUNDING_UP = 3;
const ROUNDING_DOWN = 0;

const App: React.FC = () => {
  const { baseCurrency, supportedRates: rates, updateDate, status } = useCurrencies();
  const [value = ["0.00", baseCurrency] as RateValue, setValue] = useState<RateValue>();

  const [targetCurrency = baseCurrency, setTargetCurrency] = useState<string>();

  if (status !== STATUS.FINISHED) {
    return <>Loading...</>;
  }

  const lastEditedCurrency = value[1];

  return (
    <AppStyled>
      <MenuSectionStyled>
        <Button onClick={() => console.log("cancel")} text="Cancel" />
        <Button onClick={() => console.log("exchange")} text="Exchange" />
      </MenuSectionStyled>
      <InputSectionStyled>
        {baseCurrency}
        <CurrencyAwareMoneyInput
          key={baseCurrency}
          currency={baseCurrency}
          from={rates[value[1]]}
          to={rates[baseCurrency]}
          isEditing={lastEditedCurrency === baseCurrency}
          onValueChange={setValue}
          rawValue={value[0]}
          roundingType={ROUNDING_UP}
        />
        {targetCurrency}
        <CurrencyAwareMoneyInput
          key={targetCurrency + "target"}
          from={rates[value[1]]}
          to={rates[targetCurrency]}
          currency={targetCurrency}
          isEditing={lastEditedCurrency === targetCurrency}
          onValueChange={setValue}
          rawValue={value[0]}
          roundingType={ROUNDING_DOWN}
        />
        
      </InputSectionStyled>
      <CurrencySelectSection>
        {Object.keys(rates).map(currency => <Button key={currency} onClick={() => setTargetCurrency(currency)} text={currency} />)}
      </CurrencySelectSection>
    </AppStyled>
  );
};

export default App;
