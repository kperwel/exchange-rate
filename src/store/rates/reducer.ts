import { RatesActionTypes, FETCH_FINISHED, FETCH_ERROR, ExchangeRates } from "./types";

export enum Status {
  INIT = "INIT",
  FINISHED = "FINISHED",
  ERROR = "ERROR"
}

export interface RatesState {
  rates: ExchangeRates;
  date: string;
  error: Error | null;
  baseCurrency: string;
  status: Status;
}

const initialState: RatesState = {
  rates: {},
  date: "",
  baseCurrency: "",
  error: null,
  status: Status.INIT
};

export default function userReducer(state = initialState, action: RatesActionTypes): RatesState {
  switch (action.type) {
    case FETCH_FINISHED:
      return {
        ...state,
        rates: {
          ...action.payload.response.rates,
          [action.payload.response.base]: 1
        },
        date: action.payload.response.date,
        baseCurrency: action.payload.response.base,
        status: Status.FINISHED
      };
    case FETCH_ERROR:
      return {
        ...state,
        status: Status.ERROR,
        error: action.payload.error
      };
    default:
      return state;
  }
}
