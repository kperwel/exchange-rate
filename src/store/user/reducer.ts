import { UserActionTypes, CURRENCY_CHANGE, VALUE_CHANGE, EXCHANGE, SWAP_CURRENCIES } from "./types";
import Big from "big.js";
import produce from "immer";
import { Currency } from "../../types";

interface Balance {
  [currency: string]: Big;
}

export enum MessageType {
  ERROR,
  WARNING,
  SUCCESS
}

export enum CurrencyType {
  SOURCE = "SOURCE",
  TARGET = "TARGET"
}

export interface UserState {
  [CurrencyType.SOURCE]: Currency;
  [CurrencyType.TARGET]: Currency;
  value: [string, Currency];
  balance: Balance;
}

const initialState: UserState = {
  [CurrencyType.SOURCE]: "EUR",
  [CurrencyType.TARGET]: "PLN",
  value: ["0.00", "EUR"],
  balance: {
    AUD: Big("1000.00"),
    BGN: Big("1000.00"),
    BRL: Big("1000.00"),
    CAD: Big("1000.00"),
    CHF: Big("1000.00"),
    CNY: Big("1000.00"),
    CZK: Big("1000.00"),
    DKK: Big("1000.00"),
    GBP: Big("1000.00"),
    HKD: Big("1000.00"),
    HRK: Big("1000.00"),
    HUF: Big("1000.00"),
    IDR: Big("1000.00"),
    ILS: Big("1000.00"),
    INR: Big("1000.00"),
    ISK: Big("1000.00"),
    JPY: Big("1000.00"),
    KRW: Big("1000.00"),
    MXN: Big("1000.00"),
    MYR: Big("1000.00"),
    NOK: Big("1000.00"),
    NZD: Big("1000.00"),
    PHP: Big("1000.00"),
    PLN: Big("1000.00"),
    RON: Big("1000.00"),
    RUB: Big("1000.00"),
    SEK: Big("1000.00"),
    SGD: Big("1000.00"),
    THB: Big("1000.00"),
    USD: Big("1000.00"),
    ZAR: Big("1000.00"),
    TRY: Big("1000.00"),
    EUR: Big("1000.00"),
  }
};

export default function userReducer(state = initialState, action: UserActionTypes): UserState {
  switch (action.type) {
    case CURRENCY_CHANGE:
      return {
        ...state,
        [action.payload.currencyType]: action.payload.currency
      };
    case VALUE_CHANGE:
      return {
        ...state,
        value: action.payload.value
      };
    case SWAP_CURRENCIES:
      return {
        ...state,
        [CurrencyType.SOURCE]: state[CurrencyType.TARGET],
        [CurrencyType.TARGET]: state[CurrencyType.SOURCE]
      };
    case EXCHANGE:
      const {
        payload: { sourceCurrency, targetCurrency, value, valueRate, sourceRate, targetRate }
      } = action;

      return produce(state, draftState => {
        draftState.balance[sourceCurrency] = draftState.balance[sourceCurrency] || Big("0");
        draftState.balance[targetCurrency] = draftState.balance[targetCurrency] || Big("0");

        const paidValue = Big(value[0])
          .div(valueRate)
          .mul(sourceRate);

        if (draftState.balance[sourceCurrency].gte(paidValue)) {
          const gotValue = Big(value[0])
            .div(valueRate)
            .mul(targetRate);
          draftState.value = ["0.00", value[1]];
          draftState.balance[sourceCurrency] = draftState.balance[sourceCurrency].sub(paidValue);
          draftState.balance[targetCurrency] = draftState.balance[targetCurrency].add(gotValue);
        }
      });
    default:
      return state;
  }
}
