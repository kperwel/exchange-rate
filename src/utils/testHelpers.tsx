import React from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";

export const provideStore = (
  children: React.ReactNode,
  mockedState: any,
  reducer = (state = mockedState) => state
) => <Provider store={createStore(reducer)}>{children}</Provider>;

export function HookWrapper(props: { hook: () => void }) {
  if (props.hook) {
    props.hook();
  }
  return <div />;
}

export const fetchMock = () =>
  new Promise<Response>(resolve =>
    resolve(({
      ok: true,
      status: 200,
      json: () => {
        return {};
      }
    } as any) as Response)
  );
