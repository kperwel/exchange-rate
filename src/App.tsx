import React from "react";
import CurrencyAwareMoneyInput from "./containers/CurrencyAwareMoneyInput";
import CurrencySelector from "./containers/CurrencySelector";
import ExchangeButton from "./containers/ExchangeButton";
import styled from "styled-components";

import {
  getSourceCurrency,
  getTargetCurrency,
  getRatesFetchStatus,
  getSourceBalance,
  getTargetBalance
} from "./store/selectors";

import { useSelector } from "react-redux";
import { CURRENCY_TYPE } from "./store/user/reducer";
import { STATUS } from "./store/rates/reducer";

const AppStyled = styled.div`
  display: flex;
  flex-direction: column;
  background: #fff;
  width: 100%;
  min-height: 100%;
  background-image: linear-gradient(120deg, #e0c3fc 0%, #8ec5fc 100%);

  justify-content: center;
  align-items: center;
`;

const App: React.FC = () => {
  const status = useSelector(getRatesFetchStatus);
  const sourceCurrency = useSelector(getSourceCurrency);
  const targetCurrency = useSelector(getTargetCurrency);

  const sourceBalance = useSelector(getSourceBalance);
  const targetBalance = useSelector(getTargetBalance);

  if (status === STATUS.ERROR) {
    return <>Error</>;
  } else if (status !== STATUS.FINISHED) {
    return <>Loading...</>;
  }

  return (
    <AppStyled>
      You have {sourceBalance} {sourceCurrency} and {targetBalance} {targetCurrency}. Pay
      <CurrencyAwareMoneyInput currencyType={CURRENCY_TYPE.SOURCE} />
      <CurrencySelector currencyType={CURRENCY_TYPE.SOURCE} />
      to receive
      <CurrencyAwareMoneyInput currencyType={CURRENCY_TYPE.TARGET} />
      <CurrencySelector currencyType={CURRENCY_TYPE.TARGET} />
      <ExchangeButton />
    </AppStyled>
  );
};

export default App;
