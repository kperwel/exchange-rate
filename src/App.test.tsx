import React from 'react';
import { mount } from "enzyme";
import Big from "big.js";

import App from './App';
import { Status } from './store/rates/reducer';
import { provideStore } from './utils/testHelpers';

const getState = (status: Status) => ({
  "user": {
    "SOURCE": "EUR",
    "TARGET": "PLN",
    "value": [
      "0.00",
      "EUR"
    ],
    "balance": {
      "EUR": Big("1000"),
      "PLN": Big("200")
    }
  },
  "rates": {
    "rates": {
      "PLN": 4.2675,
      "EUR": 1
    },
    "date": "2019-07-10",
    "baseCurrency": "EUR",
    "error": null,
    "status": status
  }
});

it('should mount without crashing', () => {
  mount(provideStore(<App />, getState(Status.FINISHED)));
});

it('should render initial state', () => {
  const wrapper = mount(provideStore(<App />, getState(Status.FINISHED)));
  expect(wrapper).toMatchSnapshot();
});

it('should render loading state', () => {
  const wrapper = mount(provideStore(<App />, getState(Status.INIT)));
  expect(wrapper).toMatchSnapshot();
});

it('should render error state', () => {
  const wrapper = mount(provideStore(<App />, getState(Status.ERROR)));
  expect(wrapper).toMatchSnapshot();
});

