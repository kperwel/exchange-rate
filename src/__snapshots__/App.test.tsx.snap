// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render error state 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <App>
    Error
  </App>
</Provider>
`;

exports[`should render initial state 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <App>
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bZQynM",
              "isStatic": true,
              "lastClassName": "gPKfBo",
              "rules": Array [
                "
  display: flex;
  font-size: 2em;
  width: 100%;
  min-height: 100%;

  align-items: center;
  justify-content: center;
  text-align: center;
  color: #47404e;
  line-height: 1.7em;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bZQynM",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-bZQynM gPKfBo"
        >
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-dnqmqq",
                    "isStatic": true,
                    "lastClassName": "fpgwsK",
                    "rules": Array [
                      "
  max-width: 1200px;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-dnqmqq",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="sc-dnqmqq fpgwsK"
              >
                You have 
                <styled.span>
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-gzVnrw",
                          "isStatic": true,
                          "lastClassName": "grgfgk",
                          "rules": Array [
                            "
  font-weight: bold;
",
                          ],
                        },
                        "displayName": "styled.span",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-gzVnrw",
                        "target": "span",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <span
                      className="sc-gzVnrw grgfgk"
                    >
                      1000.00
                    </span>
                  </StyledComponent>
                </styled.span>
                 
                EUR
                 and
                 
                <styled.span>
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-gzVnrw",
                          "isStatic": true,
                          "lastClassName": "grgfgk",
                          "rules": Array [
                            "
  font-weight: bold;
",
                          ],
                        },
                        "displayName": "styled.span",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-gzVnrw",
                        "target": "span",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <span
                      className="sc-gzVnrw grgfgk"
                    >
                      200.00
                    </span>
                  </StyledComponent>
                </styled.span>
                 
                PLN
                .
                <br />
                Pay 
                <CurrencyAwareMoneyInput
                  currencyType="SOURCE"
                >
                  <MoneyInput
                    onValueChange={[Function]}
                    value="0.00"
                  >
                    <styled.input
                      inputMode="numeric"
                      onChange={[Function]}
                      type="text"
                      value="0.00"
                    >
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-htpNat",
                              "isStatic": true,
                              "lastClassName": "giPlqj",
                              "rules": Array [
                                "
  font-size: inherit;
  color: inherit;
  background: #FFFFFF33;
  border: 0;
  max-width: 200px;
  margin-left: 5px;
  margin-right: 5px;
",
                              ],
                            },
                            "displayName": "styled.input",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-htpNat",
                            "target": "input",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        inputMode="numeric"
                        onChange={[Function]}
                        type="text"
                        value="0.00"
                      >
                        <input
                          className="sc-htpNat giPlqj"
                          inputMode="numeric"
                          onChange={[Function]}
                          type="text"
                          value="0.00"
                        />
                      </StyledComponent>
                    </styled.input>
                  </MoneyInput>
                </CurrencyAwareMoneyInput>
                <CurrencySelector
                  currencyType="SOURCE"
                >
                  <CurrencySelector
                    currencies={
                      Array [
                        "PLN",
                        "EUR",
                      ]
                    }
                    currency="EUR"
                    onChange={[Function]}
                  >
                    <styled.select
                      onChange={[Function]}
                      value="EUR"
                    >
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bdVaJa",
                              "isStatic": true,
                              "lastClassName": "bwiEOO",
                              "rules": Array [
                                "
  font-size: inherit;
  color: inherit;
  background: #FFFFFF33;
  border: 0;
  margin-left: 5px;
  margin-right: 5px;
",
                              ],
                            },
                            "displayName": "styled.select",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bdVaJa",
                            "target": "select",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onChange={[Function]}
                        value="EUR"
                      >
                        <select
                          className="sc-bdVaJa bwiEOO"
                          onChange={[Function]}
                          value="EUR"
                        >
                          <option
                            key="PLN"
                            value="PLN"
                          >
                            PLN
                          </option>
                          <option
                            key="EUR"
                            value="EUR"
                          >
                            EUR
                          </option>
                        </select>
                      </StyledComponent>
                    </styled.select>
                  </CurrencySelector>
                </CurrencySelector>
                to get
                <CurrencyAwareMoneyInput
                  currencyType="TARGET"
                >
                  <MoneyInput
                    onValueChange={[Function]}
                    value="0.00"
                  >
                    <styled.input
                      inputMode="numeric"
                      onChange={[Function]}
                      type="text"
                      value="0.00"
                    >
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-htpNat",
                              "isStatic": true,
                              "lastClassName": "giPlqj",
                              "rules": Array [
                                "
  font-size: inherit;
  color: inherit;
  background: #FFFFFF33;
  border: 0;
  max-width: 200px;
  margin-left: 5px;
  margin-right: 5px;
",
                              ],
                            },
                            "displayName": "styled.input",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-htpNat",
                            "target": "input",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        inputMode="numeric"
                        onChange={[Function]}
                        type="text"
                        value="0.00"
                      >
                        <input
                          className="sc-htpNat giPlqj"
                          inputMode="numeric"
                          onChange={[Function]}
                          type="text"
                          value="0.00"
                        />
                      </StyledComponent>
                    </styled.input>
                  </MoneyInput>
                </CurrencyAwareMoneyInput>
                <CurrencySelector
                  currencyType="TARGET"
                >
                  <CurrencySelector
                    currencies={
                      Array [
                        "PLN",
                        "EUR",
                      ]
                    }
                    currency="PLN"
                    onChange={[Function]}
                  >
                    <styled.select
                      onChange={[Function]}
                      value="PLN"
                    >
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bdVaJa",
                              "isStatic": true,
                              "lastClassName": "bwiEOO",
                              "rules": Array [
                                "
  font-size: inherit;
  color: inherit;
  background: #FFFFFF33;
  border: 0;
  margin-left: 5px;
  margin-right: 5px;
",
                              ],
                            },
                            "displayName": "styled.select",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bdVaJa",
                            "target": "select",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        onChange={[Function]}
                        value="PLN"
                      >
                        <select
                          className="sc-bdVaJa bwiEOO"
                          onChange={[Function]}
                          value="PLN"
                        >
                          <option
                            key="PLN"
                            value="PLN"
                          >
                            PLN
                          </option>
                          <option
                            key="EUR"
                            value="EUR"
                          >
                            EUR
                          </option>
                        </select>
                      </StyledComponent>
                    </styled.select>
                  </CurrencySelector>
                </CurrencySelector>
                <styled.span>
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-htoDjs",
                          "isStatic": true,
                          "lastClassName": "jNAluP",
                          "rules": Array [
                            "
  font-style: italic;
  display: block;
  margin-top: 0.5em;
  font-size: 0.7em;
  line-height: 1.3em;
",
                          ],
                        },
                        "displayName": "styled.span",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-htoDjs",
                        "target": "span",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <span
                      className="sc-htoDjs jNAluP"
                    >
                      You will get 
                      <TargetCurrencyPrice>
                        4.2675
                         
                      </TargetCurrencyPrice>
                       
                      PLN
                       for each 
                      EUR
                    </span>
                  </StyledComponent>
                </styled.span>
                <styled.div>
                  <StyledComponent
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-iwsKbI",
                          "isStatic": true,
                          "lastClassName": "eJYxnY",
                          "rules": Array [
                            "
  display: flex;
  flex-direction: column;
  margin: 10px;
  line-height: 1em;
",
                          ],
                        },
                        "displayName": "styled.div",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-iwsKbI",
                        "target": "div",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <div
                      className="sc-iwsKbI eJYxnY"
                    >
                      <ExchangeButton>
                        <Button
                          disabled={true}
                          message="Cannot exchange 0.00"
                          onClick={[Function]}
                          primary={true}
                        >
                          <styled.button
                            data-message="Cannot exchange 0.00"
                            disabled={true}
                            onClick={[Function]}
                            primary={true}
                            type="button"
                          >
                            <StyledComponent
                              data-message="Cannot exchange 0.00"
                              disabled={true}
                              forwardedComponent={
                                Object {
                                  "$$typeof": Symbol(react.forward_ref),
                                  "attrs": Array [],
                                  "componentStyle": ComponentStyle {
                                    "componentId": "sc-bwzfXH",
                                    "isStatic": false,
                                    "lastClassName": "dAIXbQ",
                                    "rules": Array [
                                      "
  border-radius: 4px;
  box-shadow: 2px 4px 10px 0px #0000004f;
  padding: 10px 20px;
  display: inline-flex;
  justify-content: space-between;
  align-items: center;
  font-size: 0.5em;
  margin: 10px auto;
  transition: 50ms transform linear;
  text-transform: uppercase;
  position: relative;

  ",
                                      [Function],
                                      "

  & > svg {
    margin-left: 10px;
  }

  ",
                                      [Function],
                                      "

  &:hover {
    cursor: pointer;
  }

  &:active {
    transform: scale(0.95);
  }
",
                                    ],
                                  },
                                  "displayName": "styled.button",
                                  "foldedComponentIds": Array [],
                                  "render": [Function],
                                  "styledComponentId": "sc-bwzfXH",
                                  "target": "button",
                                  "toString": [Function],
                                  "warnTooManyClasses": [Function],
                                  "withComponent": [Function],
                                }
                              }
                              forwardedRef={null}
                              onClick={[Function]}
                              primary={true}
                              type="button"
                            >
                              <button
                                className="sc-bwzfXH fMwcoR"
                                data-message="Cannot exchange 0.00"
                                disabled={true}
                                onClick={[Function]}
                                type="button"
                              >
                                Echange 
                                <TrendingUp
                                  color="currentColor"
                                  size="24"
                                >
                                  <svg
                                    fill="none"
                                    height="24"
                                    stroke="currentColor"
                                    strokeLinecap="round"
                                    strokeLinejoin="round"
                                    strokeWidth="2"
                                    viewBox="0 0 24 24"
                                    width="24"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <polyline
                                      points="23 6 13.5 15.5 8.5 10.5 1 18"
                                    />
                                    <polyline
                                      points="17 6 23 6 23 12"
                                    />
                                  </svg>
                                </TrendingUp>
                              </button>
                            </StyledComponent>
                          </styled.button>
                        </Button>
                      </ExchangeButton>
                       or 
                      <SwapButton>
                        <Button
                          onClick={[Function]}
                        >
                          <styled.button
                            onClick={[Function]}
                            type="button"
                          >
                            <StyledComponent
                              forwardedComponent={
                                Object {
                                  "$$typeof": Symbol(react.forward_ref),
                                  "attrs": Array [],
                                  "componentStyle": ComponentStyle {
                                    "componentId": "sc-bwzfXH",
                                    "isStatic": false,
                                    "lastClassName": "dAIXbQ",
                                    "rules": Array [
                                      "
  border-radius: 4px;
  box-shadow: 2px 4px 10px 0px #0000004f;
  padding: 10px 20px;
  display: inline-flex;
  justify-content: space-between;
  align-items: center;
  font-size: 0.5em;
  margin: 10px auto;
  transition: 50ms transform linear;
  text-transform: uppercase;
  position: relative;

  ",
                                      [Function],
                                      "

  & > svg {
    margin-left: 10px;
  }

  ",
                                      [Function],
                                      "

  &:hover {
    cursor: pointer;
  }

  &:active {
    transform: scale(0.95);
  }
",
                                    ],
                                  },
                                  "displayName": "styled.button",
                                  "foldedComponentIds": Array [],
                                  "render": [Function],
                                  "styledComponentId": "sc-bwzfXH",
                                  "target": "button",
                                  "toString": [Function],
                                  "warnTooManyClasses": [Function],
                                  "withComponent": [Function],
                                }
                              }
                              forwardedRef={null}
                              onClick={[Function]}
                              type="button"
                            >
                              <button
                                className="sc-bwzfXH dAIXbQ"
                                onClick={[Function]}
                                type="button"
                              >
                                Swap Currencies
                                <RefreshCw
                                  color="currentColor"
                                  size="24"
                                >
                                  <svg
                                    fill="none"
                                    height="24"
                                    stroke="currentColor"
                                    strokeLinecap="round"
                                    strokeLinejoin="round"
                                    strokeWidth="2"
                                    viewBox="0 0 24 24"
                                    width="24"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <polyline
                                      points="23 4 23 10 17 10"
                                    />
                                    <polyline
                                      points="1 20 1 14 7 14"
                                    />
                                    <path
                                      d="M3.51 9a9 9 0 0 1 14.85-3.36L23 10M1 14l4.64 4.36A9 9 0 0 0 20.49 15"
                                    />
                                  </svg>
                                </RefreshCw>
                              </button>
                            </StyledComponent>
                          </styled.button>
                        </Button>
                      </SwapButton>
                    </div>
                  </StyledComponent>
                </styled.div>
              </div>
            </StyledComponent>
          </styled.div>
        </div>
      </StyledComponent>
    </styled.div>
  </App>
</Provider>
`;

exports[`should render loading state 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <App>
    <Loader>
      <styled.div>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bxivhb",
                "isStatic": true,
                "lastClassName": "frJinh",
                "rules": Array [
                  "
  height: 100%;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bxivhb",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="sc-bxivhb frJinh"
          >
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-ifAKCX",
                      "isStatic": true,
                      "lastClassName": "jLWndV",
                      "rules": Array [
                        "
  animation: ",
                        Keyframes {
                          "id": "sc-keyframes-iVXCSc",
                          "inject": [Function],
                          "name": "iVXCSc",
                          "rules": Array [
                            "@-webkit-keyframes iVXCSc{from{-webkit-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}to{-webkit-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}",
                            "@keyframes iVXCSc{from{-webkit-transform:rotate(0deg);-ms-transform:rotate(0deg);transform:rotate(0deg);}to{-webkit-transform:rotate(360deg);-ms-transform:rotate(360deg);transform:rotate(360deg);}}",
                          ],
                          "toString": [Function],
                        },
                        " 2s linear infinite;
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-ifAKCX",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-ifAKCX jLWndV"
                >
                  <Loader
                    color="currentColor"
                    size={32}
                  >
                    <svg
                      fill="none"
                      height={32}
                      stroke="currentColor"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth="2"
                      viewBox="0 0 24 24"
                      width={32}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <line
                        x1="12"
                        x2="12"
                        y1="2"
                        y2="6"
                      />
                      <line
                        x1="12"
                        x2="12"
                        y1="18"
                        y2="22"
                      />
                      <line
                        x1="4.93"
                        x2="7.76"
                        y1="4.93"
                        y2="7.76"
                      />
                      <line
                        x1="16.24"
                        x2="19.07"
                        y1="16.24"
                        y2="19.07"
                      />
                      <line
                        x1="2"
                        x2="6"
                        y1="12"
                        y2="12"
                      />
                      <line
                        x1="18"
                        x2="22"
                        y1="12"
                        y2="12"
                      />
                      <line
                        x1="4.93"
                        x2="7.76"
                        y1="19.07"
                        y2="16.24"
                      />
                      <line
                        x1="16.24"
                        x2="19.07"
                        y1="7.76"
                        y2="4.93"
                      />
                    </svg>
                  </Loader>
                </div>
              </StyledComponent>
            </styled.div>
            <styled.span>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-EHOje",
                      "isStatic": true,
                      "lastClassName": "eSNhqD",
                      "rules": Array [
                        "
  margin-left: 10px;
",
                      ],
                    },
                    "displayName": "styled.span",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-EHOje",
                    "target": "span",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <span
                  className="sc-EHOje eSNhqD"
                >
                  Loading...
                </span>
              </StyledComponent>
            </styled.span>
          </div>
        </StyledComponent>
      </styled.div>
    </Loader>
  </App>
</Provider>
`;
